<a href="https://colab.research.google.com/github/JoseAugustoVital/Decision-Score-MarketPlace/blob/main/decision_score.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ***UNIVERSIDADE FEDERAL DO MATO GROSSO DO SUL***
# Análise de dados para aumentar nível de satisfação dos clientes de um marketplace utilizando árvore de decisão.
**TRABALHO 3 - INTELIGÊNCIA ARTIFICIAL 2021/1**
__________________________________________________

**Aluno:** 

Nome: **José Augusto Lajo Vieira Vital** 



In [ ]:
# INÍCIO DO ESTUDO

**Importação do sistema operacional**

In [ ]:
import os

**Determinação do acesso ao diretório**

In [ ]:
os.chdir('/content/drive/MyDrive/datasets')

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
!pwd

/content/drive/My Drive/datasets


In [ ]:
!ls

olist_customers_dataset.csv	  olist_products_dataset.csv
olist_geolocation_dataset.csv	  olist_sellers_dataset.csv
olist_order_items_dataset.csv	  product_category_name_translation.csv
olist_order_payments_dataset.csv  tabela_resulta.csv
olist_order_reviews_dataset.csv   tabela_result.csv
olist_orders_dataset.csv


**Importação das bibliotecas Pandas e Numpy**

In [ ]:
import pandas as pd
import numpy as np

**Leitura dos tabelas .csv**

In [ ]:
tabela_cliente = pd.read_csv('olist_customers_dataset.csv')
tabela_localizacao = pd.read_csv('olist_geolocation_dataset.csv')
tabela_pedido = pd.read_csv('olist_order_items_dataset.csv')
tabela_pagamento = pd.read_csv('olist_order_payments_dataset.csv')
tabela_review = pd.read_csv('olist_order_reviews_dataset.csv')
tabela_entrega_pedido = pd.read_csv('olist_orders_dataset.csv')
tabela_descricao_produto = pd.read_csv('olist_products_dataset.csv')
tabela_vendedor = pd.read_csv('olist_sellers_dataset.csv')
tabela_categoria_traduzido = pd.read_csv('product_category_name_translation.csv')

**Checagem dos 5 primeiros elementos de cada tabela**

In [ ]:
tabela_cliente.head()

customer_id  ... customer_state
0  06b8999e2fba1a1fbc88172c00ba8bc7  ...             SP
1  18955e83d337fd6b2def6b18a428ac77  ...             SP
2  4e7b3e00288586ebd08712fdd0374a03  ...             SP
3  b2b6027bc5c5109e529d4dc6358b12c3  ...             SP
4  4f2d8ab171c80ec8364f7c12e35b23ad  ...             SP

[5 rows x 5 columns]

In [ ]:
tabela_localizacao.head()

geolocation_zip_code_prefix  ...  geolocation_state
0                         1037  ...                 SP
1                         1046  ...                 SP
2                         1046  ...                 SP
3                         1041  ...                 SP
4                         1035  ...                 SP

[5 rows x 5 columns]

In [ ]:
tabela_pedido.head()

order_id  order_item_id  ...   price freight_value
0  00010242fe8c5a6d1ba2dd792cb16214              1  ...   58.90         13.29
1  00018f77f2f0320c557190d7a144bdd3              1  ...  239.90         19.93
2  000229ec398224ef6ca0657da4fc703e              1  ...  199.00         17.87
3  00024acbcdf0a6daa1e931b038114c75              1  ...   12.99         12.79
4  00042b26cf59d7ce69dfabb4e55b4fd9              1  ...  199.90         18.14

[5 rows x 7 columns]

In [ ]:
tabela_pagamento.head()

order_id  ...  payment_value
0  b81ef226f3fe1789b1e8b2acac839d17  ...          99.33
1  a9810da82917af2d9aefd1278f1dcfa0  ...          24.39
2  25e8ea4e93396b6fa0d3dd708e76c1bd  ...          65.71
3  ba78997921bbcdc1373bb41e913ab953  ...         107.78
4  42fdf880ba16b47b59251dd489d4441a  ...         128.45

[5 rows x 5 columns]

In [ ]:
tabela_review.head()

review_id  ... review_answer_timestamp
0  7bc2406110b926393aa56f80a40eba40  ...     2018-01-18 21:46:59
1  80e641a11e56f04c1ad469d5645fdfde  ...     2018-03-11 03:05:13
2  228ce5500dc1d8e020d8d1322874b6f0  ...     2018-02-18 14:36:24
3  e64fb393e7b32834bb789ff8bb30750e  ...     2017-04-21 22:02:06
4  f7c4243c7fe1938f181bec41a392bdeb  ...     2018-03-02 10:26:53

[5 rows x 7 columns]

In [ ]:
tabela_entrega_pedido.head()

order_id  ... order_estimated_delivery_date
0  e481f51cbdc54678b7cc49136f2d6af7  ...           2017-10-18 00:00:00
1  53cdb2fc8bc7dce0b6741e2150273451  ...           2018-08-13 00:00:00
2  47770eb9100c2d0c44946d9cf07ec65d  ...           2018-09-04 00:00:00
3  949d5b44dbf5de918fe9c16f97b45f8a  ...           2017-12-15 00:00:00
4  ad21c59c0840e6cb83a9ceb5573f8159  ...           2018-02-26 00:00:00

[5 rows x 8 columns]

In [ ]:
tabela_descricao_produto.head()

product_id  ... product_width_cm
0  1e9e8ef04dbcff4541ed26657ea517e5  ...             14.0
1  3aa071139cb16b67ca9e5dea641aaa2f  ...             20.0
2  96bd76ec8810374ed1b65e291975717f  ...             15.0
3  cef67bcfe19066a932b7673e239eb23d  ...             26.0
4  9dc1a7de274444849c219cff195d0b71  ...             13.0

[5 rows x 9 columns]

In [ ]:
tabela_vendedor.head()

seller_id  ...  seller_state
0  3442f8959a84dea7ee197c632cb2df15  ...            SP
1  d1b65fc7debc3361ea86b5f14c68d2e2  ...            SP
2  ce3ad9de960102d0677a81f5d0bb7b2d  ...            RJ
3  c0f3eea2e14555b6faeea3dd58c1b1c3  ...            SP
4  51a04a8a6bdcb23deccc82b0b80742cf  ...            SP

[5 rows x 4 columns]

In [ ]:
tabela_categoria_traduzido.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

**Início do processo de união das 9 tabelas disponibilizadas com a finalidade de produzir uma tabela resultante que possua os elementos mais importantes para a determinação do review_score. No primero merge realizado, unimos a tabela de clientes com as respectivas entregas dos pedidos usando o código individual de cada consumidor como parâmetro.**

In [ ]:
pd.merge(tabela_cliente, tabela_entrega_pedido, on=["customer_id"], how="inner")

customer_id  ... order_estimated_delivery_date
0      06b8999e2fba1a1fbc88172c00ba8bc7  ...           2017-06-05 00:00:00
1      18955e83d337fd6b2def6b18a428ac77  ...           2018-02-06 00:00:00
2      4e7b3e00288586ebd08712fdd0374a03  ...           2018-06-13 00:00:00
3      b2b6027bc5c5109e529d4dc6358b12c3  ...           2018-04-10 00:00:00
4      4f2d8ab171c80ec8364f7c12e35b23ad  ...           2018-08-15 00:00:00
...                                 ...  ...                           ...
99436  17ddf5dd5d51696bb3d7c6291687be6f  ...           2018-04-25 00:00:00
99437  e7b71a9017aa05c9a7fd292d714858e8  ...           2018-04-20 00:00:00
99438  5e28dfe12db7fb50a4b2f691faecea5e  ...           2018-05-02 00:00:00
99439  56b18e2166679b8a959d72dd06da27f9  ...           2017-12-05 00:00:00
99440  274fa6071e5e17fe303b9748641082c8  ...           2018-01-08 00:00:00

[99441 rows x 12 columns]

**Processo de união com as demais tabelas disponibilizadas**

**1 - (Clientes, Entregas)**

**2 - (1, Pedidos)**

**3 - (2, Pagamentos)**

**4 - (3, Review)**

**5 - (4, Vendedor)**

In [ ]:
test = pd.merge(tabela_cliente, tabela_entrega_pedido, on=["customer_id"], how="inner")
test = pd.merge(test, tabela_pedido, on=["order_id"], how="inner")
test = pd.merge(test, tabela_pagamento, on=["order_id"], how="inner")
test = pd.merge(test, tabela_review, on=["order_id"], how="inner")
test = pd.merge(test, tabela_vendedor, on=["seller_id"], how="inner")

**Tabela Resultante**

**Linhas: 118315**

**Colunas: 31**

In [ ]:
test

customer_id  ... seller_state
0       06b8999e2fba1a1fbc88172c00ba8bc7  ...           SP
1       4e7b3e00288586ebd08712fdd0374a03  ...           SP
2       b2b6027bc5c5109e529d4dc6358b12c3  ...           SP
3       4c06b42fbf7b97ab10779cda5549cd1c  ...           SP
4       8276de07ef25225d412b8462d73f8664  ...           SP
...                                  ...  ...          ...
118310  eda10d7eb51a675d000956fa1aabc7dd  ...           PR
118311  f0f671d4034e98cdf20f0c452d6db02b  ...           SP
118312  b90b857a277b080d8078332b3636ab58  ...           PR
118313  1ed0c832c2dd99570a59260e71768bdf  ...           SP
118314  1ed0c832c2dd99570a59260e71768bdf  ...           SP

[118315 rows x 31 columns]

**Segunda filtragem consiste em remover elementos que não possuem
relação com a variável review_score**

In [ ]:
#test = test.drop(columns=["customer_unique_id"],axis=1)
#test = test.drop(columns=["customer_city"],axis=1)
#test = test.drop(columns=["customer_state"],axis=1)
#test = test.drop(columns=["order_status"],axis=1)
#test = test.drop(columns=["order_purchase_timestamp"],axis=1)
#test = test.drop(columns=["order_approved_at"],axis=1)
#test = test.drop(columns=["order_delivered_carrier_date"],axis=1)
#test = test.drop(columns=["order_delivered_customer_date"],axis=1)
#test = test.drop(columns=["order_estimated_delivery_date"],axis=1)
#test = test.drop(columns=["shipping_limit_date"],axis=1)
#test = test.drop(columns=["review_creation_date"],axis=1)
#test = test.drop(columns=["review_answer_timestamp"],axis=1)
#test = test.drop(columns=["seller_city"],axis=1)
#test = test.drop(columns=["seller_state"],axis=1)
#test = test.drop(columns=["review_comment_title"],axis=1)
#test = test.drop(columns=["review_comment_message"],axis=1)

**Tabela Resultante após a remoção de atributos não prioritários para o nível de
satisfação dos clientes**

**Linhas: 118315**

**Colunas: 15**

In [ ]:
test

customer_id  ...  seller_zip_code_prefix
0       06b8999e2fba1a1fbc88172c00ba8bc7  ...                    8577
1       4e7b3e00288586ebd08712fdd0374a03  ...                    8577
2       b2b6027bc5c5109e529d4dc6358b12c3  ...                    8577
3       4c06b42fbf7b97ab10779cda5549cd1c  ...                    8577
4       8276de07ef25225d412b8462d73f8664  ...                    8577
...                                  ...  ...                     ...
118310  eda10d7eb51a675d000956fa1aabc7dd  ...                   87015
118311  f0f671d4034e98cdf20f0c452d6db02b  ...                    9861
118312  b90b857a277b080d8078332b3636ab58  ...                   83015
118313  1ed0c832c2dd99570a59260e71768bdf  ...                   12914
118314  1ed0c832c2dd99570a59260e71768bdf  ...                   12914

[118315 rows x 15 columns]

**Inserindo cada atributo da tabela resultante em um vetor para melhor manipulação dos dados**


In [ ]:
vetor_cliente = np.array(test.customer_id)
vetor_cepcliente = np.array(test.customer_zip_code_prefix)
vetor_pedido = np.array(test.order_id)
vetor_idpedido = np.array(test.order_item_id)
vetor_produto = np.array(test.product_id)
vetor_vendedor = np.array(test.seller_id)
vetor_preco_produto = np.array(test.price)
vetor_frete = np.array(test.freight_value)
vetor_parcela = np.array(test.payment_sequential)
vetor_tipopagamento = np.array(test.payment_type)
vetor_pay = np.array(test.payment_installments)
vetor_valorfinal = np.array(test.payment_value)
vetor_review = np.array(test.review_id)
vetor_score = np.array(test.review_score)
vetor_cepvendedor = np.array(test.seller_zip_code_prefix)

**Definindo um novo dataframe vazio**

In [ ]:
df = pd.DataFrame()
df

Empty DataFrame
Columns: []
Index: []

**Definindo as colunas do novo dataframe e atribuindo para cada coluna, seu respectivo vetor de dados registrado anteriormente.**

In [ ]:
COLUNAS = [
    'Cliente',
    'CEP_Cliente',
    'Pedido',
    'id_Pedido',
    'Produto',
    'Vendedor',
    'Preco_produto',
    'Frete',
    'Parcela',
    'Tipo_pagamento',
    'Installments',
    'Valor_total',
    'ID_Review',
    'CEP_Vendedor',
    'Score'
]

df = pd.DataFrame(columns =COLUNAS)

df.Cliente = vetor_cliente
df.CEP_Cliente = vetor_cepcliente
df.Pedido = vetor_pedido
df.id_Pedido = vetor_idpedido
df.Produto = vetor_produto
df.Vendedor = vetor_vendedor
df.Preco_produto = vetor_preco_produto
df.Frete = vetor_frete
df.Parcela = vetor_parcela
df.Tipo_pagamento = vetor_tipopagamento
df.Installments = vetor_pay
df.Valor_total = vetor_valorfinal
df.ID_Review = vetor_review
df.CEP_Vendedor = vetor_cepvendedor
df.Score = vetor_score
df

Cliente  CEP_Cliente  ... CEP_Vendedor  Score
0       06b8999e2fba1a1fbc88172c00ba8bc7        14409  ...         8577      4
1       4e7b3e00288586ebd08712fdd0374a03         1151  ...         8577      5
2       b2b6027bc5c5109e529d4dc6358b12c3         8775  ...         8577      5
3       4c06b42fbf7b97ab10779cda5549cd1c        65075  ...         8577      1
4       8276de07ef25225d412b8462d73f8664        90010  ...         8577      4
...                                  ...          ...  ...          ...    ...
118310  eda10d7eb51a675d000956fa1aabc7dd        39920  ...        87015      5
118311  f0f671d4034e98cdf20f0c452d6db02b         7231  ...         9861      3
118312  b90b857a277b080d8078332b3636ab58        76400  ...        83015      5
118313  1ed0c832c2dd99570a59260e71768bdf        37610  ...        12914      1
118314  1ed0c832c2dd99570a59260e71768bdf        37610  ...        12914      1

[118315 rows x 15 columns]

**Impressão da coluna de clientes.**

In [ ]:
df.Cliente

0         06b8999e2fba1a1fbc88172c00ba8bc7
1         4e7b3e00288586ebd08712fdd0374a03
2         b2b6027bc5c5109e529d4dc6358b12c3
3         4c06b42fbf7b97ab10779cda5549cd1c
4         8276de07ef25225d412b8462d73f8664
                        ...               
118310    eda10d7eb51a675d000956fa1aabc7dd
118311    f0f671d4034e98cdf20f0c452d6db02b
118312    b90b857a277b080d8078332b3636ab58
118313    1ed0c832c2dd99570a59260e71768bdf
118314    1ed0c832c2dd99570a59260e71768bdf
Name: Cliente, Length: 118315, dtype: object

In [ ]:
#for index, row in df.iterrows():
#    if row['Score'] == 1:
#        df.loc[index,'Classe'] =  'Pessimo'
#    if row['Score'] == 2:
#        df.loc[index,'Classe'] =  'Ruim'
#    if row['Score'] == 3:
#        df.loc[index,'Classe'] =  'Mediano'
#    if row['Score'] == 4:
#        df.loc[index,'Classe'] =  'Bom'
#    if row['Score'] == 5:
#        df.loc[index,'Classe'] =  'Otimo'

**Informações do dataframe**

**Atributos, elementos não nulos, Tipo das variáveis da coluna**

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118315 entries, 0 to 118314
Data columns (total 15 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Cliente         118315 non-null  object 
 1   CEP_Cliente     118315 non-null  int64  
 2   Pedido          118315 non-null  object 
 3   id_Pedido       118315 non-null  int64  
 4   Produto         118315 non-null  object 
 5   Vendedor        118315 non-null  object 
 6   Preco_produto   118315 non-null  float64
 7   Frete           118315 non-null  float64
 8   Parcela         118315 non-null  int64  
 9   Tipo_pagamento  118315 non-null  object 
 10  Installments    118315 non-null  int64  
 11  Valor_total     118315 non-null  float64
 12  ID_Review       118315 non-null  object 
 13  CEP_Vendedor    118315 non-null  int64  
 14  Score           118315 non-null  int64  
dtypes: float64(3), int64(6), object(6)
memory usage: 13.5+ MB


**Agrupando os elementos do dataframe por consumidor**

In [ ]:
df.groupby(by='Cliente').size()

Cliente
00012a2ce6f8dcda20d059ce98491703    1
000161a058600d5901f007fab4c27140    1
0001fd6190edaaf884bcaf3d49edf079    1
0002414f95344307404f0ace7a26f1d5    1
000379cdec625522490c315e70c7a9fb    1
                                   ..
fffcb937e9dd47a13f05ecb8290f4d3e    1
fffecc9f79fd8c764f843e9951b11341    3
fffeda5b6d849fbd39689bb92087f431    1
ffff42319e9b2d713724ae527742af25    1
ffffa3172527f765de70084a7e53aae8    2
Length: 98665, dtype: int64

**Importando os métodos de árvore de decisão**

In [ ]:
from sklearn.tree import DecisionTreeClassifier, export_graphviz
from sklearn.model_selection import train_test_split
from sklearn import metrics

**Para simplificar os dados e evitar criar um dummy com esse dataframe, removemos todos os elementos não-numéricos para que o modelo seja capaz de realizar a execução. Solução encontrada para simplificar os atributos do tipo "Objeto" para tipos numéricos.**

In [ ]:
df['Cliente'] = df['Cliente'].str.replace(r'\D', '')
df['Pedido'] = df['Pedido'].str.replace(r'\D', '')
df['Produto'] = df['Produto'].str.replace(r'\D', '')
df['Vendedor'] = df['Vendedor'].str.replace(r'\D', '')
df['ID_Review'] = df['ID_Review'].str.replace(r'\D', '')
df

Cliente  CEP_Cliente  ... CEP_Vendedor  Score
0             068999211881720087        14409  ...         8577      4
1         4730028858608712037403         1151  ...         8577      5
2          260275510952946358123         8775  ...         8577      5
3             406427971077955491        65075  ...         8577      1
4       827607252254128462738664        90010  ...         8577      4
...                          ...          ...  ...          ...    ...
118310          1075167500095617        39920  ...        87015      5
118311       0671403498200452602         7231  ...         9861      3
118312  908572770808078332363658        76400  ...        83015      5
118313     108322995705926071768        37610  ...        12914      1
118314     108322995705926071768        37610  ...        12914      1

[118315 rows x 15 columns]

**Realizamos o procedimento de remoção dos elementos não-numéricos para todas as colunas do tipo objeto com exceção do tipo de pagamento pois o tipo de pagamento se resume a poucas opções. Dessa forma, usamos a função get_dummies apenas para o tipo de pagamento**

**Portanto, a coluna Tipo_pagamento se divide em quatro colunas com lógica booleana. As novas colunas são: Tipo_pagamento_boleto, 	Tipo_pagamento_credit_card, Tipo_pagamento_debit_card, Tipo_pagamento_voucher**

In [ ]:
result_df = pd.get_dummies(df, columns=["Tipo_pagamento"])


**Resultado final do dataframe**

In [ ]:
result_df

Cliente  ...  Tipo_pagamento_voucher
0             068999211881720087  ...                       0
1         4730028858608712037403  ...                       0
2          260275510952946358123  ...                       0
3             406427971077955491  ...                       0
4       827607252254128462738664  ...                       0
...                          ...  ...                     ...
118310          1075167500095617  ...                       0
118311       0671403498200452602  ...                       0
118312  908572770808078332363658  ...                       0
118313     108322995705926071768  ...                       0
118314     108322995705926071768  ...                       0

[118315 rows x 18 columns]

**Criação de um dataframe reserva para possíveis conclusões**

In [ ]:
reserva = result_df
reserva

Cliente  ...  Tipo_pagamento_voucher
0             068999211881720087  ...                       0
1         4730028858608712037403  ...                       0
2          260275510952946358123  ...                       0
3             406427971077955491  ...                       0
4       827607252254128462738664  ...                       0
...                          ...  ...                     ...
118310          1075167500095617  ...                       0
118311       0671403498200452602  ...                       0
118312  908572770808078332363658  ...                       0
118313     108322995705926071768  ...                       0
118314     108322995705926071768  ...                       0

[118315 rows x 18 columns]

**Eliminando todas as linhas com nível 4 ou nível 5 de satisfação. Dessa forma, temos um dataframe com todos os dados, e um apenas com dados classificados com nível 3,2 ou 1 ou seja, mediano, ruim ou péssimo. (elementos que apresentam nível de insatisfação interessante para análise)**

In [ ]:
reserva = reserva.drop(reserva[reserva.Score > 3].index)
reserva


Cliente  ...  Tipo_pagamento_voucher
3             406427971077955491  ...                       0
7         0372810041857927016263  ...                       0
18             45800571938861445  ...                       0
19          19530467435951219042  ...                       0
21             52814168460899591  ...                       0
...                          ...  ...                     ...
118306  049671902463228846651580  ...                       0
118309     211018165721598633401  ...                       0
118311       0671403498200452602  ...                       0
118313     108322995705926071768  ...                       0
118314     108322995705926071768  ...                       0

[29593 rows x 18 columns]

**Processo de separação de treine/teste**

**Proporções estabelecidas:**

**70% treino**

**30% teste**

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(result_df.drop('Score', axis=1), result_df['Score'], test_size=0.3)

**Número de amostras para cada processo**

In [ ]:
X_train.shape, X_test.shape

((82820, 17), (35495, 17))

**Número de targets para cada processo**

In [ ]:
y_train.shape, y_test.shape

((82820,), (35495,))

**Criação do classifcador**

In [ ]:
cls = DecisionTreeClassifier()

**Treinamento**

In [ ]:
cls = cls.fit(X_train, y_train)

**Vetor com as importancias de cada atributo para a determinação do review_score**

In [ ]:
cls.feature_importances_

array([0.10390396, 0.11024563, 0.10356182, 0.00756018, 0.09445855,
       0.07974123, 0.08029503, 0.08109755, 0.00294333, 0.03654096,
       0.08643976, 0.11395647, 0.07748421, 0.0083304 , 0.00813804,
       0.00214091, 0.00316196])

In [ ]:
df.head()

Cliente  CEP_Cliente  ... CEP_Vendedor  Score
0        068999211881720087        14409  ...         8577      4
1    4730028858608712037403         1151  ...         8577      5
2     260275510952946358123         8775  ...         8577      5
3        406427971077955491        65075  ...         8577      1
4  827607252254128462738664        90010  ...         8577      4

[5 rows x 15 columns]

**Para tornar o modelo mais visual, criamos um laço para a impressão dos pesos de cada atributo para determinar o score**

In [ ]:
for feature, importancia in zip(result_df.columns, cls.feature_importances_):
    print("{}:{:.1f}%".format(feature,((importancia*100))))


Cliente:10.4%
CEP_Cliente:11.0%
Pedido:10.4%
id_Pedido:0.8%
Produto:9.4%
Vendedor:8.0%
Preco_produto:8.0%
Frete:8.1%
Parcela:0.3%
Installments:3.7%
Valor_total:8.6%
ID_Review:11.4%
CEP_Vendedor:7.7%
Score:0.8%
Tipo_pagamento_boleto:0.8%
Tipo_pagamento_credit_card:0.2%
Tipo_pagamento_debit_card:0.3%


**Vetor com predições para checagem do aprendizado**

In [ ]:
result = cls.predict(X_test)
result

array([1, 4, 5, ..., 5, 5, 4])

In [ ]:
result_df.Score[118310]

5

**Representação das métricas de precisão e médias do modelo**

In [ ]:
from sklearn import metrics
print(metrics.classification_report(y_test,result))

              precision    recall  f1-score   support

           1       0.26      0.27      0.27      4602
           2       0.12      0.13      0.12      1193
           3       0.15      0.17      0.16      2975
           4       0.24      0.25      0.24      6685
           5       0.61      0.58      0.60     20040

    accuracy                           0.43     35495
   macro avg       0.28      0.28      0.28     35495
weighted avg       0.44      0.43      0.44     35495



**Precisão total**

In [ ]:
from sklearn.model_selection import cross_val_score
allScores = cross_val_score(cls, X_train, y_train , cv=10)
allScores.mean()

0.42534411977783143

**Treinamento utilizando o dataframe reserva. (apenas com os níveis de satisfação abaixo da média, score <=3)** 

**Split do dataframe em treino e teste (70%, 30%, respectivamente)**

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(reserva.drop('Score', axis=1), reserva['Score'], test_size=0.3)

**Quantidade de amostras do treino**

In [ ]:
X_train.shape, X_test.shape


((20715, 17), (8878, 17))

**Classificador**

In [ ]:
clf = DecisionTreeClassifier()

**Treino**

In [ ]:
clf = clf.fit(X_train, y_train)


**Importancia de cada atributo para determinar o nível de satisfação dos consumidores**

In [ ]:
clf.feature_importances_

array([0.10048426, 0.11080533, 0.10888151, 0.00838036, 0.09478009,
       0.08306931, 0.0741649 , 0.08334586, 0.00208028, 0.03551857,
       0.08657828, 0.11647515, 0.07639348, 0.00679141, 0.0070284 ,
       0.00174576, 0.00347703])

In [ ]:
for feature, importancia in zip(reserva.columns, clf.feature_importances_):
    print("{}:{:.1f}%".format(feature,((importancia*100))))
    

Cliente:10.0%
CEP_Cliente:11.1%
Pedido:10.9%
id_Pedido:0.8%
Produto:9.5%
Vendedor:8.3%
Preco_produto:7.4%
Frete:8.3%
Parcela:0.2%
Installments:3.6%
Valor_total:8.7%
ID_Review:11.6%
CEP_Vendedor:7.6%
Score:0.7%
Tipo_pagamento_boleto:0.7%
Tipo_pagamento_credit_card:0.2%
Tipo_pagamento_debit_card:0.3%


# ANÁLISE DOS DADOS PROCESSADOS

**Com o objetivo de aumentar o review_score desse marketplace, utilizamos o algoritmo de árvore de decisão para encontrar elementos que influenciam diretamente a aceitação e satisfação dos clientes, analisando desde o processo de despacho do produto, até a qualidade do atendimento e do produto final. Dessa forma, utilizamos estratégias de filtragem. Inicialmente eliminamos os elementos que, evidentemente, não apresentavam influencia sobre a nota registrada pelo cliente. Após esse passo inicial, simplificamos o tipo dos dados para facilitar o processamento do dataframe pelo algoritmo. Após o processo de preparar os dados para o aprendizado registramos os seguintes valores:**

**Considerando todos as notas (Ótimo, bom, mediano, ruim, péssimo) :**

Cliente:10.4%

CEP_Cliente:11.0%

Pedido:10.4%

id_Pedido:0.8%

Produto:9.4%

Vendedor:8.0%

Preco_produto:8.0%

Frete:8.1%

Parcela:0.3%

Installments:3.7%

Valor_total:8.6%

ID_Review:11.4%

CEP_Vendedor:7.7%

Score:0.8%

Tipo_pagamento_boleto:0.8%

Tipo_pagamento_credit_card:0.2%

Tipo_pagamento_debit_card:0.3%

------------------------------------

**Considerando apenas avaliações com notas instatisfatórias (mediano, ruim, péssimo) :**

Cliente:10.0%

CEP_Cliente:11.1%

Pedido:10.9%

id_Pedido:0.8%

Produto:9.5%

Vendedor:8.3%

Preco_produto:7.4%

Frete:8.3%

Parcela:0.2%

Installments:3.6%

Valor_total:8.7%

ID_Review:11.6%

CEP_Vendedor:7.6%

Score:0.7%

Tipo_pagamento_boleto:0.7%

Tipo_pagamento_credit_card:0.2%

Tipo_pagamento_debit_card:0.3%

---------------------
**Portanto, com base nos resultados obtidos, o dono do marketplace deve se atentar aos seguintes parâmetros de sua logística:**

**1 - Relação entre CEP_Cliente, Frete e CEP_Vendedor**

**Esses atributos apresentaram importancia direta na nota registrada pelo cliente, dessa forma, algumas razões devem ser consideradas:**

**-Problemas com a entrega.**

**-Qualidade da entrega(tempo de entrega,comprometimento do produto durante o processo de transporte).**

**-Alto preço do frete para determinadas regiões.**

-----------------------------
**2 - Produto**

**O atributo produto apresentou importancia direta na insatisfação dos clientes. Portanto, deve-se considerar:**

**-Qualidade ruim de determinados produtos.**

**-O produto entregue não apresentar as caracteristicas do produto anunciado.**

**-Entrega de produtos errados, problema de logística.**

---------------------------------

**3 - Vendedor**

**O atributo Vendedor apresentou importancia direta na insatisfação dos clientes. Portanto, deve-se considerar:**

**-Qualidade duvidosa do atendimento por parte do vendedor.**

**-Erro do vendedor em alguma etapa do processo em específico.**

--------------------------


# **CONCLUSÃO**


**Portanto, é possível concluir que para o marketplace aumentar seu score, analisar problemas logísticos por parte do processo de transporte, aspectos do produto e os vendedores são os fatores mais importantes para entender o que se passa na empresa e solucionar o problema que está gerando insatisfação dos clientes desse e-commerce.**





